frontend/catalyst/cuda/__init__.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains a CudaQDevice and the qjit
+entry point.
+"""
+
+from pathlib import Path
+
+import cudaq
+import pennylane as qml
+
+from catalyst.cuda.catalyst_to_cuda_interpreter import interpret
+
+
+def qjit(fn=None, **kwargs):
+    """Wrapper around QJIT for CUDA-quantum."""
+
+    if fn is not None:
+        return interpret(fn)
+
+    def wrap_fn(fn):
+        return interpret(fn)
+
+    return wrap_fn
+
+
+class BaseCudaInstructionSet(qml.QubitDevice):
+    """Base instruction set for CUDA-Quantum devices"""
+
+    # TODO: Once 0.35 is released, remove -dev suffix.
+    pennylane_requires = "0.35.0-dev"
+    version = "0.1.0"
+    author = "Xanadu, Inc."
+
+    # pylint: disable=duplicate-code
+    operations = [
+        "CNOT",
+        "CY",
+        "CZ",
+        "CRX",
+        "CRY",
+        "CRZ",
+        "PauliX",
+        "PauliY",
+        "PauliZ",
+        "Hadamard",
+        "S",
+        "T",
+        "RX",
+        "RY",
+        "RZ",
+        "SWAP",
+        # "CSWAP", This is a bug in cuda-quantum. CSWAP is not exposed.
+    ]
+    observables = []
+    config = Path(__file__).parent / "cuda_quantum.toml"
+
+    def __init__(self, shots=None, wires=None, mps=False, multi_gpu=False):
+        self.mps = mps
+        self.multi_gpu = multi_gpu
+        super().__init__(wires=wires, shots=shots)
```

Comment:
This is a very minor comment, so not needed for approval of this PR. But `multi_gpu` and `mps` are both options specific only to `NvidiaCuStateVec` and `NvidiaCuTensorNet` respectively, so it would make more sense for them to be defined in the subclasses. E.g.,

```python
class NvidiaCuStateVec(BaseCudaInstructionSet):
    """Concrete device class for CuStateVec"""

    name = "CuStateVec"
    short_name = "nvidia.custatevec"

    def __init__(self, shots=None, wires=None, multi_gpu=False):
        self.multi_gpu = multi_gpu
        return super().__init__(shots=None, wires=None)
```

That is, the base class should only include options that are universal

Reply:
This is fine, but since I cannot test this devices yet I don't want to make them available yet. Can we leave this to a soon to follow PR where we can also test these ones?

Reply:
@erick-xanadu recommend making this small change, just because it means `multi_gpu` and `mps` are options to `softwareq.qpp`, which doesn't make sense 🤔